### PR TITLE
Permission constraints

### LINKED ISSUE #593: Permission popup: add constraints
## Overview
Continue the work start on the permission popup: 

Some permissions are depending on other permissions, for instance, you cannot give "can_grant_view:content" to a user without giving him "can_view:content" first.

Implement them all in a proper way (readable code).

image

## Acceptance criteria
- [ ] Implement the rules described in the [devdoc perm page](https://france-ioi.github.io/algorea-devdoc/design/access-rights/items/#changing-and-propagating-permissions), in the "Constraint on “receiver”" column.
In practice, in the UI:
- [ ] do not let the user reduce the "required" permission, by greying out the lower option (and its icon on the left) and displaying an explanation tooltip on hover
- [ ] do not let the user increases the "requiring" permission,  by greying out the lower option (and its icon on the left) and displaying an explanation tooltip on hover

### PR DESCRIPTION
**Pull request by GabrielVidal1**
(this PR still needs some work, I just need some insights)

Fixes #593 

~~The non valid options in the permissions edition dialog box are now grayed out according to the [constraints on receiver](https://france-ioi.github.io/algorea-devdoc/design/access-rights/items/#changing-and-propagating-permissions).~~

~~The only missing behavior is when a permission is changed that invalidates another, the invalid one is set to the closest valid one For example if someone has `can_view = content` and `can_watch = result`, if we set `can_view` to `info` the receiver cannot have `can_watch = result` because he needs `can_view ≥ content`, so the `can_watch` permissions should be also updated.~~

 ~~My solution would be to make the `ProgressSectionValue` and `BooleanSectionComponent` components implement  `ControlValueAccessor` to make the permission edition dialog a form. It will be easier to manipulate and changes can be handled directly from the `PermissionsEditDialogComponent`. The permissions would also be saved if changed (if the form is dirty)~~

~~The only issue is that without strongly typed forms a lot of problems can occur when getting the value from the form and when setting them programmatically.
I used a packaged that made forms strongly typed in another PR (#472) and I can really see how it could be used here to simplify everything form related~~

See [here](https://github.com/France-ioi/AlgoreaFrontend/pull/692#issuecomment-940177540) for the update

### CODE REVIEW

**Comment by smadbe** [comment]
I am not available this week. @ThomasJuster , could you have a look at Gabriel's question?

**Comment by ThomasJuster** [comment]
I'll try addressing every matter, if I miss one please notify me :pray: 

&nbsp;

## When a permission change invalidates another

> The only missing behavior is when a permission is changed that invalidates another, the invalid one is set to the closest valid one

Good point, however I think automatically setting the invalid value(s) to the closest one is not really ok because the user might not see we made the change for him/her.

An alternative would be to make use of form validation with a custom validator. That way the user would notified of the incorrect values and take action on his/her own. This obviously advocates for implementing `ControlValueAccessor` as well :smile: 

&nbsp;

## Implement `ControlValueAccessor` − integrate with angular forms

> My solution would be to make the ProgressSectionValue and BooleanSectionComponent components implement ControlValueAccessor to make the permission edition dialog a form. It will be easier to manipulate and changes can be handled directly from the PermissionsEditDialogComponent. The permissions would also be saved if changed (if the form is dirty)

I think those components should indeed implement `ControlValueAccessor`, but if we are doing that, we should go further and factorise the current code:
Technically, instead of one `ProgressSectionValue` and one `BooleanSectionComponent`, what we should have here is:
1. A collapse section component, which is only an interactive layout component and has nothing to do with forms
2. A select with a special UI (progress section value) which should implement `ControlValueAccessor`
3. A switch which already exists, already implements ControlValueAccessor, and should allow [projecting](https://angular.io/guide/content-projection) a label

<details>
  <summary>What it would give in terms of code when using the components (click me):</summary>

```html
Before:
<alg-progress-section>…</alg-progress-section>
<alg-boolean-section>…</alg-boolean-section>

After: (component names are just for the example)
<alg-collapse-section>
  <alg-progress-select>…</alg-progress-select>
 </alg-collapse-section>

<alg-collapse-section>
  <alg-boolean-select>…</alg-boolean-select>
 </alg-collapse-section>
```

</details>

Now going down this road induces a lot of changes, I would suggest to split this work and make multiple small PRs. Your call.

In my opinion we should go this way, however if you don't have time for such changes, we should stick to make existing `*-section` components implement `ControlValueAccessor`

&nbsp;

## Possible issues because forms are not strongly typed

> The only issue is that without strongly typed forms a lot of problems can occur when getting the value from the form and when setting them programmatically.
I used a packaged that made forms strongly typed in another PR (#472) and I can really see how it could be used here to simplify everything form related

In my opinion, although having types is a good idea, this can also be (partially) handled in the components implementing `ControlValueAccessor`

Considering we don't have many forms, I don't know if installing `@ngneat/reactive-forms` makes sense, I think I don't have authority for this decision, it will have to be approved by @smadbe.
No matter the final decision, installing or not the package is not a blocking to finish the issue, to me it can be arbitrated later.

&nbsp;

**Footnotes:**

That comment is quite long with a lot of information and opinions, don't hesitate to ping me for a chat or a call :relaxed:

**Comment by GabrielVidal1** [comment]
After refactoring a lot of the code (mainly `collapsible-section`, `progress-select` and `switch-select`), permissions constraints are finally added through a custom validator in the permissions edition dialog box as suggested [here](https://github.com/France-ioi/AlgoreaFrontend/pull/692#issuecomment-925890649).

I still have some questions.

## `is_owner=true` and `can_make_session_official=true`
I guessed if the user we're changing permissions already have `is_owner=true` (or `can_make_session_official=true`) then the constraints doesn't apply, so the field isn't valid only if the value was false and set to true (and the requirements are not met).
Is this the right way? 

## UI
I added a `validationErrors` input to the `progress-select` and `switch-select` components, so that they display the error (which is the list of the requirements that are not met).
### Valid value
image
* The lock icon is opened
### Invalid value
image
* The errors are displayed next to an exclamation triangle icon below the `progress-select`
* The lock is locked
### Error messages
For now error messages are temporary but I tried to make them convey the requirements that are not met. Is there a specific error format I should follow?  
## i18n
A lot of the error messages share some parts (for example, the 'You need at least' or 'This user needs').
Should these string be split into multiple strings? or should each error should be kept whole?

**Comment by smadbe** [comment]
> I still have some questions.
> ## `is_owner=true` and `can_make_session_official=true`
> 
> I guessed if the user we're changing permissions already have `is_owner=true` (or `can_make_session_official=true`) then the constraints doesn't apply, so the field isn't valid only if the value was false and set to true (and the requirements are not met). Is this the right way?

I am not sure I fully understood your question.  You should be able to change `is_owner=true` to `false (same for `can_make_session_official`) without specific constraints on the "receiving user".

> ## UI
> 
> I added a `validationErrors` input to the `progress-select` and `switch-select` components, so that they display the error (which is the list of the requirements that are not met).
> ### Valid value
> 
> image
> 
>     * The lock icon is opened
> 
> 
> ### Invalid value
> 
> image
> 
>     * The errors are displayed next to an exclamation triangle icon below the `progress-select`
> 
>     * The lock is locked

I don't think using the lock/unlock icon make things clear.  I think it was even proposed to remove the lock icon which is misleading. 
Shouldn't you grey out the "solution" selector if it is invalid and write (even if solution is not selected) in dark red your constraint just after (on a new line) or instead) "The user can also see the solution ...".

> ### Error messages
> 
> For now error messages are temporary but I tried to make them convey the requirements that are not met. Is there a specific error format I should follow?

There are no specific format, it should be clear enough for a usual user, which may not be very easy.

> ## i18n
> 
> A lot of the error messages share some parts (for example, the 'You need at least' or 'This user needs'). Should these string be split into multiple strings? or should each error should be kept whole?

Cannot you do this nested expressions https://angular.io/guide/i18n-common-prepare#mark-alternates-and-nested-expressions ?

File: src/app/modules/item/helpers/item-permissions-constraints.ts [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+): { isOwner?: string[] } {
+
+  if (receiverPermissions.isOwner && !previousValue && !giverPermissions.isOwner) {
+    return { isOwner: [ youNeedToBeOwnerText ] };
+  }
+  return {};
+}
+
+export function permissionsConstraintsValidator(
+  receiverPermissions: GroupPermissions,
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.ts:185]: This is to handle the change for `is_owner` and `can_make_session_official`.

In the following explication, the giver is the user changing the permissions, and the receiver is the user whose permissions are being changed.
To my understanding, the receiver being an owner of the item (permission maybe granted by someone else) doesn't necessarily mean the giver is one, so the form would be invalidated right away, because the current giver doesn't have the permission to grant `is_owner` to the receiver.

That's why I need to have the initial value of the form, so there is an exception (the aforementioned case, receiver is already an owner, but not the giver) and only test the case "receiver goes from not-owner to owner"

I'm not 100% sure of this edge case, but that's what I deducted from the permissions constraints in the [devdoc](https://france-ioi.github.io/algorea-devdoc/design/access-rights/items/#changing-and-propagating-permissions) for the "is_owner=true" granted perm.

File: src/app/modules/item/helpers/item-permissions-constraints.ts [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+import { TypeFilter } from '../components/composition-filter/composition-filter.component';
+import { PermissionsDialogData } from '../components/permissions-edit-dialog/permissions-edit-dialog.component';
+import {
+  generateCanViewValues,
+  generateCanGrantViewValues,
+  generateCanWatchValues,
+  generateCanEditValues
+} from '../components/permissions-edit-dialog/permissions-edit-dialog-texts';
+
+const youNeedToBeOwnerText = $localize`You need to be <b>owner</b> of this item`;
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.ts:13]: After some checking, this is doable, but not a good practice :relieved: so I got rid of the html tags in localized strings

File: src/app/modules/item/helpers/item-permissions-constraints.ts [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+const youNeedText = $localize`You need`;
+const thisUserNeedsText = $localize`This user needs`;
+const toBeText = $localize`to be`;
+const toBeAtLeastText = $localize`to be at least`;
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.ts:17]: fixed, I also got rid of the html tags for good measures

File: src/app/modules/item/helpers/item-permissions-constraints.ts [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+function atLeast<T extends readonly string[]>(values: T, val: T[number], ref: T[number]): boolean {
+  return values.indexOf(val) >= values.indexOf(ref);
+}
+
+function validateCanView(receiverPermissions: GroupPermissions, giverPermissions: PermissionsInfo): { canView?: string[] } {
+
+  if (receiverPermissions.canView === 'info' &&
+  !atLeast(canGrantViewValues, giverPermissions.canGrantView, 'enter')) {
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.ts:26]: fixed, indeed this is much more readable

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html [hunk 5]
```diff
@@ -22,12 +22,16 @@
...
+          i18n-title title="Can view" icon="fa fa-eye"
+          [collapsible]="!form.errors?.canView"
+        >
           <ng-template #content let-collapsed>
             <alg-progress-select
               [collapsed]="collapsed"
               formControlName="canView"
-              [values]="canViewValues"
+              [values]="progressSelectValues.canViewValues"
+              [validationErrors]="form.errors?.canView"
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html:34]: Because the validation depends not on a single form control but on multiple (for example, the `can_grant_view` permission depends on the `can_view`), so the custom validator concerns the form group, and not individual form controls.

I tried to create a custom validator for each of the fields, but getting the value of adjacent fields (by getting the parent and then the form control of the other fields) but it seems like a bad practice, and it made the code more complex overall.

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts [hunk 6]
```diff
@@ -37,25 +47,21 @@ export class PermissionsEditDialogComponent implements OnChanges {
...
-        canWatch: this.permissions.can_watch,
-        canEdit: this.permissions.can_edit,
-        canMakeSessionOfficial: this.permissions.can_make_session_official,
-        isOwner: this.permissions.is_owner,
-      }, { emitEvent: false });
+  ngOnChanges(_changes: SimpleChanges): void {
+    if (this.permissions && this.giverPermissions) {
+      this.form.setValidators(permissionsConstraintsValidator(this.permissions, this.giverPermissions));
+      this.form.updateValueAndValidity();
+      this.form.reset({ ...this.permissions }, { emitEvent: false });
```
[the remark this answers is not in the document]
Comment (GabrielVidal1) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.ts:64]: fixed, although the `targetType` is needed when generating the text displayed for the different permissions, so I included it as inputs that refresh the progress-select values

File: src/app/modules/item/helpers/item-permissions-constraints.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/modules/item/helpers/item-permissions-constraints.ts]: Nice :+1: 
I think there's the same issue in [this file](https://github.com/France-ioi/AlgoreaFrontend/pull/692/files#diff-0de36960acf645f141e7281609465a4855801d913c403212186e513c4ad45813R25), the injected sentence subject is English-specific

File: src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/modules/item/components/permissions-edit-dialog/permissions-edit-dialog.component.html:34]: I don't understand the link with validation, to me validation is defined at form-level and errors are stored into `form.errors` object, into properties that may match input names

Here, input name and error property seem to match, so according to the [doc](https://angular.io/api/forms/AbstractControl#errors) we should be able to also access the error from the `canView` FormControl via property `errors`

File: src/app/modules/item/helpers/item-permissions-constraints.ts [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    errors.push(`This user needs ${bolden(permissionsInfoString.canView.string)} to be at least ${
+      bolden(permissionsInfoString.canView.content)}`);
+  }
+
+  if (receiverPermissions.canWatch === 'answer_with_grant' && !giverPermissions.isOwner) {
+    errors.push($localize`You need to be owner of this item`);
+  } else {
+
+    // if receiverPermissions.canWatch is 'result' or 'answer'
+    if (!(giverPermissions.canWatch === 'answer_with_grant')) {
```
Comment (ThomasJuster) [inline src/app/modules/item/helpers/item-permissions-constraints.ts:122]: Mmh :arrow_right: how about using … `!==` ? :smile:

Comment (GabrielVidal1) [reply src/app/modules/item/helpers/item-permissions-constraints.ts]: fixed

**Comment by GabrielVidal1** [comment]
This PR has been continued [here](https://github.com/France-ioi/AlgoreaFrontend/pull/881)
